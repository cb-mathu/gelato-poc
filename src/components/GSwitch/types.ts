import { VueComponentProps } from '../../types';
import { ButtonHTMLAttributes, ExtractPropTypes } from 'vue';
import { SWITCH_PROPS } from './constants';

export type SwitchProps = ExtractPropTypes<typeof SWITCH_PROPS>;
export type NativeButtonProps = Omit<ButtonHTMLAttributes, keyof SwitchProps>;
export type MergedSwitchProps = Partial<SwitchProps & VueComponentProps & NativeButtonProps>;
