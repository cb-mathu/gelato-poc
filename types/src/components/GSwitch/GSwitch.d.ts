import { MergedSwitchProps } from './types';
declare const _default: new () => {
    $props: MergedSwitchProps;
};
export default _default;
