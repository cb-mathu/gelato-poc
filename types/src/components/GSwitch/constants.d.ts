import { PropType } from 'vue';
export declare const SWITCH_PROPS: {
    label: {
        type: StringConstructor;
    };
    modelValue: {
        type: BooleanConstructor;
    };
    'onUpdate:modelValue': PropType<(value: boolean) => void>;
    id: {
        type: StringConstructor;
        required: boolean;
    };
    class: {
        type: StringConstructor;
        required: boolean;
        default: string;
    };
    disabled: {
        type: BooleanConstructor;
    };
    autofocus: {
        type: BooleanConstructor;
        default: boolean;
    };
    autocomplete: {
        type: StringConstructor;
        default: string;
    };
    required: {
        type: BooleanConstructor;
        default: boolean;
    };
    readonly: {
        type: BooleanConstructor;
        default: boolean;
    };
    name: {
        type: StringConstructor;
        default: string;
    };
    placeholder: {
        type: StringConstructor;
        default: string;
    };
    modelModifiers: {
        default: () => {};
    };
    invalid: {
        type: BooleanConstructor;
    };
    errorMessage: {
        type: StringConstructor;
    };
};
