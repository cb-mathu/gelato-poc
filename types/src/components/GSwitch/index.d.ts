import GSwitch from './GSwitch';
export { GSwitch };
