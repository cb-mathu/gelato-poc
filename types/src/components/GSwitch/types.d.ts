import { VueComponentProps } from '../../types';
import { ButtonHTMLAttributes, ExtractPropTypes } from 'vue';
import { SWITCH_PROPS } from './constants';
export declare type SwitchProps = ExtractPropTypes<typeof SWITCH_PROPS>;
export declare type NativeButtonProps = Omit<ButtonHTMLAttributes, keyof SwitchProps>;
export declare type MergedSwitchProps = Partial<SwitchProps & VueComponentProps & NativeButtonProps>;
