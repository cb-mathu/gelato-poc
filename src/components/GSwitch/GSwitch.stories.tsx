import { ref } from 'vue';
import GSwitch from './GSwitch';
import { GNotification } from '../GNotification';
import { MergedSwitchProps } from './types';

export default {
  title: 'Component/Forms/GSwitch',
  component: GSwitch,
  argTypes: {
    disabled: {
      description: 'Disable the field',
      control: {
        type: 'boolean',
      },
    },
  },
};

const Template = (args: any) => ({
  setup() {
    const toggleEnabled = ref(false);
    return () => (
      <div style="display:grid;grid-gap: 0.75rem;">
        <GSwitch {...args} v-model={toggleEnabled.value} id="input" />
        <GNotification style="margin-top: 1rem">
          Toggled :<b>{toggleEnabled.value ? 'On' : 'Off'}</b>
        </GNotification>
      </div>
    );
  },
});

const slots = { label: () => <span>Enable</span> };

export const Story = Template.bind({});
export const DefaultSwitch = (props?: MergedSwitchProps) => <GSwitch {...(props as any)} />;
export const DisabledOnSwitch = () => <GSwitch modelValue={true} disabled />;
export const DisabledOffSwitch = () => <GSwitch modelValue={false} disabled />;
export const SwitchWithLabel = (props?: MergedSwitchProps) => <GSwitch {...(props as any)} label="Enable" />;
export const SwitchWithLabelSlot = () => <GSwitch v-slots={slots} />;
export const InvalidSwitch = () => <GSwitch v-slots={slots} invalid errorMessage="This is an error message, ooh poor you!" />;
