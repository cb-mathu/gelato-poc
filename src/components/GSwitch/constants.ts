import { PropType } from 'vue';
import { BASE_FORM_PROPS } from '../../utils/props';

export const SWITCH_PROPS = {
  ...BASE_FORM_PROPS,
  label: {
    type: String,
  },
  modelValue: {
    type: Boolean,
  },
  'onUpdate:modelValue': Function as PropType<(value: boolean) => void>,
};
