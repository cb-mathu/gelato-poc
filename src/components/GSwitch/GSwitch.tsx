import { defineComponent, ref, watchEffect } from 'vue';
import { SWITCH_PROPS } from './constants';
import { Switch, SwitchGroup, SwitchLabel } from '@headlessui/vue';
import { MergedSwitchProps, SwitchProps } from './types';
import { GInputErrorMessage } from '../GInput';
import { useFormAriaIds } from '../../utils/accessibility';
import { isNotEmpty } from '../../utils/common';

const GSwitch = defineComponent({
  name: 'GSwitch',
  props: SWITCH_PROPS,
  emits: ['update:modelValue'],
  setup(props: SwitchProps, { slots, attrs }) {
    let internalValue = ref(props.modelValue);
    watchEffect(() => {
      internalValue = ref(props.modelValue);
    });
    return () => {
      const label = props.label || slots.label?.();
      const { errorId } = useFormAriaIds(props.id);
      const attributes = {
        ...attrs,
        ...props,
      };

      const classes = {
        'g-switch': true,
        'g-switch--active': internalValue.value,
        'g-switch--disabled': props.disabled,
        'g-switch--invalid': props.invalid,
      };

      return (
        <SwitchGroup as="div">
          <div class={classes}>
            <Switch {...attributes} v-model={internalValue.value} as="button" class="g-switch__element">
              <span class="sr-only">Use setting</span>
              <span aria-hidden="true" class="g-switch__control" />
            </Switch>
            {isNotEmpty(label) && <SwitchLabel class="g-switch__label">{label}</SwitchLabel>}
          </div>
          {props.invalid && <GInputErrorMessage id={errorId} errorMessage={props.errorMessage} />}
        </SwitchGroup>
      );
    };
  },
});

export default GSwitch as {
  new (): {
    $props: MergedSwitchProps;
  };
};
