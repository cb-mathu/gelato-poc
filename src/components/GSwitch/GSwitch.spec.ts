import { render, fireEvent } from '@testing-library/vue';
import { screen } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';
import * as stories from './GSwitch.stories';

const { DefaultSwitch, DisabledOffSwitch, DisabledOnSwitch, SwitchWithLabel, SwitchWithLabelSlot } = stories;

describe('GSwitch : ', () => {
  it('renders switch by default', () => {
    const { container } = render(DefaultSwitch());
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders button tag by default', () => {
    const { container } = render(DefaultSwitch());
    const buttonElement = container.querySelector('button');
    expect(buttonElement).toBeTruthy();
  });

  it('renders label passed via prop', () => {
    render(SwitchWithLabel());
    expect(screen.getByText('Enable')).toBeTruthy();
  });

  it('renders custom label passed via slot', () => {
    render(SwitchWithLabelSlot());
    expect(screen.getByText('Enable')).toBeTruthy();
  });

  it('updates the state on click', async () => {
    render(SwitchWithLabel({ modelValue: false }));
    await fireEvent.click(screen.getByText('Enable'));
    expect(screen.getByRole('switch').getAttribute('aria-checked')).toBe('true');
  });

  it('pressing tab should focus the switch', () => {
    render(SwitchWithLabel());
    userEvent.tab();
    expect(screen.getByRole('switch')).toHaveFocus();
  });

  it('pressing tab should not focus disabled switch', () => {
    render(DisabledOffSwitch());
    userEvent.tab();
    expect(screen.getByRole('switch')).not.toHaveFocus();
  });

  it('emits the updated state (v-model) on click', async () => {
    const handleClick = jest.fn();
    render(SwitchWithLabel({ modelValue: false, 'onUpdate:modelValue': handleClick }));
    await fireEvent.click(screen.getByText('Enable'));
    expect(handleClick).toHaveBeenCalledWith(true);
  });

  it('renders on switch in disabled state', () => {
    const { container } = render(DisabledOnSwitch());
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders off switch in disabled state', () => {
    const { container } = render(DisabledOffSwitch());
    expect(container.firstChild).toMatchSnapshot();
  });
});
