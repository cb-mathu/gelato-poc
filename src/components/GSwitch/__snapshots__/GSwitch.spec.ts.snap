// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GSwitch :  renders off switch in disabled state 1`] = `
<div>
  <div
    class="g-switch g-switch--disabled"
  >
    <button
      aria-checked="false"
      autocomplete="off"
      class="g-switch__element"
      disabled=""
      id="headlessui-switch-15"
      invalid="false"
      modelmodifiers="[object Object]"
      modelvalue="false"
      name=""
      placeholder=""
      required="false"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        class="sr-only"
      >
        Use setting
      </span>
      <span
        aria-hidden="true"
        class="g-switch__control"
      />
    </button>
    <!---->
  </div>
  <!---->
</div>
`;

exports[`GSwitch :  renders on switch in disabled state 1`] = `
<div>
  <div
    class="g-switch g-switch--active g-switch--disabled"
  >
    <button
      aria-checked="true"
      autocomplete="off"
      class="g-switch__element"
      disabled=""
      id="headlessui-switch-14"
      invalid="false"
      modelmodifiers="[object Object]"
      modelvalue="true"
      name=""
      placeholder=""
      required="false"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        class="sr-only"
      >
        Use setting
      </span>
      <span
        aria-hidden="true"
        class="g-switch__control"
      />
    </button>
    <!---->
  </div>
  <!---->
</div>
`;

exports[`GSwitch :  renders switch by default 1`] = `
<div>
  <div
    class="g-switch"
  >
    <button
      aria-checked="false"
      autocomplete="off"
      class="g-switch__element"
      id="headlessui-switch-1"
      invalid="false"
      modelmodifiers="[object Object]"
      modelvalue="false"
      name=""
      placeholder=""
      required="false"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        class="sr-only"
      >
        Use setting
      </span>
      <span
        aria-hidden="true"
        class="g-switch__control"
      />
    </button>
    <!---->
  </div>
  <!---->
</div>
`;
